### PR TITLE
add return_indices to max_pool2d and implement max_unpool2d

### PR DESCRIPTION
**Pull request by geohotstan**

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -2073,7 +2074,27 @@ def max_pool2d(self, kernel_size=(2,2), stride=None, dilation=1, padding=0, ceil
...
     pads = self._ceil_mode_padding2d(k_, stride if stride is not None else k_, dilation, padding) if ceil_mode else self._padding2d(padding, len(k_))
-    return self.pad(pads, value=dtypes.min(self.dtype))._pool(k_, stride if stride is not None else k_, dilation).max(tuple(range(-len(k_), 0)))
+    pooled = self.pad(pads, value=dtypes.min(self.dtype))._pool(k_, stride if stride is not None else k_, dilation)
+    ret = pooled.max(tuple(range(-len(k_), 0)))
+    if not return_indices: return ret
+    # HACK: add +1 offset to indices for `masked = (ret == flat_pool).where(flat_indices, 0)` to work with no 0's in `flat_indices`
+    indices = Tensor.arange(1, self.numel() + 1, dtype=self.dtype, device=self.device, requires_grad=False).reshape(self.shape)
+    indices = indices.pad(pads, value=dtypes.min(self.dtype))._pool(k_, stride if stride is not None else k_, dilation)
+    flat_indices, flat_pool = indices.flatten(-len(k_)), pooled.flatten(-len(k_))
+    # TODO: this contiguous changes result
```
Comment (geohotstan) [inline tinygrad/tensor.py:2084]: this implementation is a mess
but with this messy implementation I found another `.contiguous()` bug
result is wrong without this `contiguous`

maybe we do need a contiguous fuzzer 🤔

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -2073,7 +2074,27 @@ def max_pool2d(self, kernel_size=(2,2), stride=None, dilation=1, padding=0, ceil
...
-    return self.pad(pads, value=dtypes.min(self.dtype))._pool(k_, stride if stride is not None else k_, dilation).max(tuple(range(-len(k_), 0)))
+    pooled = self.pad(pads, value=dtypes.min(self.dtype))._pool(k_, stride if stride is not None else k_, dilation)
+    ret = pooled.max(tuple(range(-len(k_), 0)))
+    if not return_indices: return ret
+    # HACK: add +1 offset to indices for `masked = (ret == flat_pool).where(flat_indices, 0)` to work with no 0's in `flat_indices`
+    indices = Tensor.arange(1, self.numel() + 1, dtype=self.dtype, device=self.device, requires_grad=False).reshape(self.shape)
+    indices = indices.pad(pads, value=dtypes.min(self.dtype))._pool(k_, stride if stride is not None else k_, dilation)
+    flat_indices, flat_pool = indices.flatten(-len(k_)), pooled.flatten(-len(k_))
+    # TODO: this contiguous before reduce-where chain changes the result, without it the results are wrong
+    masked = (ret.unsqueeze(-1) == flat_pool).where(flat_indices, 0).contiguous()
```
Comment (geohotstan) [inline tinygrad/tensor.py:2085]: this implementation is a mess
but with this messy implementation I found another `.contiguous()` bug
result is wrong without this `contiguous`

maybe we do need a contiguous fuzzer 🤔

posting this PR early to see if anyone have any ideas why this is bugged.

Comment (chenyuxyz) [reply tinygrad/tensor.py:2085]: i can take a look if you have a smaller repro

Comment (geohotstan) [reply tinygrad/tensor.py:2085]: > i can take a look if you have a smaller repro

okok let me try to minimize it.

Comment (geohotstan) [reply tinygrad/tensor.py:2085]: 5dd61035f7e68b82998411b04b99d32bb1e6259c
this revert fixed it 👀

**Comment by geohotstan** [comment]
probably going to have the correct implementation left in onnx_ops

will do this is #8109

**Pull request closed without merging**